docs/_posts/2022-11-09-time-aware-tiered-storage.markdown

```diff
@@ -0,0 +1,121 @@
+
+Compare to normal compaction which only outputs the data to a single level, Per-key placement compaction can output data to 2 different levels, as per-per placement compaction is only for the last level compaction, so the 2 output levels would **always be the penultimate level, and the last level**. The compaction places the key to its corresponding tier by simply checking the key’s sequence number.
+
+At the beginning of the compaction, the compaction job collects all seq to time information from every input SSTs and merges them together, then based on the current time to get the oldest sequence number that should be put into non-last-level (hot tier). During the last level compaction, as long as the key is newer than the oldest_sequence_number, it will be placed in the penultimate level (hot tier) instead of the last level (cold tier).
+
+Note, RocksDB also places the keys that are within the user snapshot in the hot tier, there’re a few reasons for that:
+1. It’s reasonable to assume snapshot-protected data are hot data;
+2. Avoid mixing the sequence number not zeroed out data with old last-level data, which is desirable to reduce the oldest obsolete data time (it’s defined as the oldest SST time that has a non-zero sequence number). It also means tombstones are always placed in the hot tier, which is also desirable as it should be pretty small.
+3. The original motivation was to avoid moving data from the lower level to a higher level in case the user increases the [`preclude_last_level_data_seconds`](https://github.com/facebook/rocksdb/blob/b0d9776b704af01c2b5385e9d53754e0c8176373/include/rocksdb/advanced_options.h#L927), so the snapshot-protected data in the last level will become hot again, and moving data to a higher level. It’s not always safe to move data from a lower level to a higher level in the LSM tree which could cause key conflict. Later we added a conflict check to allow the data to move up as long as there’s no key conflict, but then the movement is not guaranteed (see Migration for details)
+
+### Migration
+Once the user enables the feature, it enables both time tracking and per-key placement compaction **at the same time**. As the existing data, it can still be mismarked as cold data. To have a smooth migration to the feature. The user can enable the time-tracking feature first. For example, if the user plans to set [`preclude_last_level_data_seconds`](https://github.com/facebook/rocksdb/blob/b0d9776b704af01c2b5385e9d53754e0c8176373/include/rocksdb/advanced_options.h#L927) to 3 days, the user can enable time tracking 3 days earlier with [`preserve_internal_time_seconds`](https://github.com/facebook/rocksdb/blob/b0d9776b704af01c2b5385e9d53754e0c8176373/include/rocksdb/advanced_options.h#L950). Then when enabling the tiered storage feature, it already has the time information for the last 3 days' hot data, then per-key placement compaction won’t compact them to the last level.
```

Comment:
> As the existing data, it can still be mismarked as cold data.

Does this mean a seqno older than all entries in the timestamp mapping always maps to cold tier?

Reply:
It's more about the existing data that are already compacted to the last level which doesn't have seqno. If a key doesn't have seqno, it always maps to the cold tier.

For a key that has seqno, it depends.
* if a seqno older than all entries and all entries are larger (later) than the `preclude_time`, it will be map to hot tier. https://github.com/facebook/rocksdb/blob/b36ec37a4b8bc6e6b17ea386c6abb29abc76900b/db/compaction/compaction_job.cc#L337-L338 preclude_last_level_min_seqno_ will be set to 0. So all data with seqno will be hot (it's conservative, but assuming the majority of the data is already in the last level with seqno zeroed).
* But even without enabling the time tracking feature, it still collects the seqno -> time from the sst metadata:
https://github.com/facebook/rocksdb/blob/b36ec37a4b8bc6e6b17ea386c6abb29abc76900b/db/compaction/compaction_job.cc#L301-L302
which is to improve the time estimation accuracy, if the time tracking map has a more accurate entry, that Add() will be ignored:
https://github.com/facebook/rocksdb/blob/c401f285c3215043346e150e6a7f319946d2f325/db/seqno_to_time_mapping.cc#L286-L290
in this case, if there's no time tracking entries, then the metadata entry will be used. So it will try to estimate the time.